docs/source/designs/backup-recovery.rst

```diff
@@ -0,0 +1,515 @@
+   A snapshot can be produced using the ``iterdump`` feature of the Python SQLite3 module.
+
+#. an event stream
+
+   An event stream is a sequence of SQL statements
+   (with arguments embedded)
+   which update a certain database snapshot.
+   Each statements in the event stream is paired with a sequence number.
+   Sequence numbers are assigned in a monotonically increasing sequence that corresponds to the order in which the statements were executed.
+   These can be used to address a specific statement within the event stream.
+
+#. a sequence number
+
+   A snapshot includes state which was created by statements from some prefix of the event stream.
+   The snapshot is paired with a sequence number indicating the last statement in this prefix.
+   This allows recovery to find the correct position in the event stream to begin replaying statements to reconstruct the newest version of the database.
+
+The event stream is represented in the local database in a new table::
+
+  CREATE TABLE [event-stream] (
+      -- A sequence number which allows us to identify specific positions in
+      -- the sequence of modifications which were made to the database.
+      [sequence-number] INTEGER PRIMARY KEY,
+
+      -- A SQL statement which likely made a change to the database state.
+      [statement] TEXT,
+  );
+
+Arguments are substituted into the statement so that they match the form of statements generated during the *snapshot* phase.
+
+Replication
+~~~~~~~~~~~
+
+The replication process is as follows:
+
+#. Replication is configured using the external interface.
+
+   #. The *replica directory*,
+      a new mutable directory,
+      is created on grid.
+   #. The write capability is added to the database.
+   #. The read capability is returned to the external caller.
+
+#. If there is not a sufficiently up-to-date snapshot [1]_ on the grid then one is created [7]_ in the *replica directory*.
+   Any obsolete snapshots [2]_ in the *replica directory* are pruned.
+
+#. As the application runs the event stream is recorded [3]_ locally in the database.
+
+#. If the event stream in the database is large enough [4]_ or the application triggers an event stream flush [5]_ then:
+
+   #. it is added to the event stream in the *replica directory* [6]_
+   #. statements which were added are pruned from the database [8]_
+
+#. If an event stream object in the *replica directory* contains only statements that are already part of the snapshot those statements are pruned. [9]_
+
+All uploads inherit the redundancy configuration from the Tahoe-LAFS client node.
+
+Recovery
+~~~~~~~~
+
+The recovery process is as follows:
+
+#. An empty database is created.
+#. The snapshot is downloaded.
+#. The event stream is downloaded.
+#. The statements from the snapshot are executed against the database.
+#. The statements from the event stream,
+   starting at the first statement after the snapshot's sequence number,
+   are executed against the database.
+
```

Comment:
To follow a previous comment above, I wonder if caps should also be rotated here to minimize the odds of multiple actors writing to the same mutable/directory.

Reply:
I assumed (but maybe this should be explicit) that a post-recovery instance would need to be told to start doing backups again..? (and then create a new writecap etc)

Reply:
Are there backup/recovery user stories/acceptance criteria already written somewhere that talks about expectations about data read/write access?

Reply:
Ignoring the user entirely: it's much easier to reason about "will we have precisely one writer to this writecap" (i.e. meet the Tahoe constraints) if a new one is created for each "instance" of zkap-authorizer (even if it's a recovery of another instance).